perfkitbenchmarker/providers/azure/azure_network.py

```diff
@@ -21,264 +21,428 @@
+    self.vnet = vnet
+    self.name = name
+    self.args = ['--vnet-subnet-name', self.name]
+
+  def _Create(self):
+    logging.info('subnet._Create')
+    vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'vnet', 'subnet', 'create',
+         '--vnet-name', self.vnet.name,
+         '--address-prefix', self.vnet.address_space,
+         self.name] + self.resource_group.args)
+
+  def _Exists(self):
+    stdout, _, retcode = vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'vnet', 'subnet', 'show',
+         '--vnet-name', self.vnet.name,
+         '--json',
+         self.name] + self.resource_group.args)
+
+    return retcode == 0 and stdout != '{}\n'
+
+  def _Delete(self):
+    vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'vnet', 'subnet', 'delete',
+         '--vnet-name', self.vnet.name,
+         '--quiet',
+         self.name] + self.resource_group.args)
+
+
+class AzureNetworkSecurityGroup(resource.BaseResource):
+  def __init__(self, location, subnet, name):
+    super(AzureNetworkSecurityGroup, self).__init__()
+
+    self.location = location
+    self.subnet = subnet
+    self.name = name
+    self.resource_group = GetResourceGroup()
+    self.args = ['--nsg-name', self.name]
+
+    self.rules_lock = threading.Lock()
+    # Mapping of port -> rule name, used to deduplicate rules. We
+    # expect duplicate rules because PKB will call AllowPort() for
+    # each VM on a subnet, but the rules are actually applied to the
+    # entire subnet.
+    self.rules = {}
+    # True if the special 'DenyAll' rule is present. If deny_all_mode
+    # is True, then rules should be empty.
+    self.deny_all_mode = False
+
+  def _Create(self):
+    vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'nsg', 'create',
+         '--location', self.location,
+         self.name] + self.resource_group.args)
+
+  def _Exists(self):
+    stdout, _, retcode = vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'nsg', 'show',
+         '--json',
+         self.name] + self.resource_group.args)
+
+    return retcode == 0 and stdout != '{}\n'
+
+  def _Delete(self):
+    vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'network', 'nsg', 'delete',
+         '--quiet',
+         self.name] + self.resource_group.args)
+
+  def AttachToSubnet(self):
+    vm_util.IssueRetryableCommand(
+        [azure.AZURE_PATH, 'network', 'vnet', 'subnet', 'set',
+         '--name', self.subnet.name,
+         '--network-security-group-name', self.name] +
+        self.resource_group.args +
+        self.subnet.vnet.args)
+
+  def DeleteRule(self, rule):
+      vm_util.IssueRetryableCommand(
+          [azure.AZURE_PATH, 'network', 'nsg', 'rule', 'delete',
+           '--quiet',
+           rule] +
+          self.resource_group.args +
+          self.args)
+
+  def AllowPort(self, vm, port):
+    with self.rules_lock:
+      if port in self.rules:
+        return
+      if self.deny_all_mode:
```

Comment:
Having `AllowPort` silently override a previous `DisallowAllPorts` makes me feel a little uneasy. My intuition would make me expect that all of the ports except this port I'm allowing should still be closed. But meh, I doubt this will cause any issues.